### PR TITLE
Setting up automated deployments

### PR DESCRIPTION
**Pull request by angelazhu2**
**features**: 
- docker file to set up the github actions and create a frontend and backend image to push to docker hub. 
- set up database secrets added to github secrets
- edited the user and password secrets to the codebase in Github settings. 
- there was an update to main, that lead to the compose file depending on .env files. This was rectified by adding database secrets to github secrets and took out the dependency in docker-compose. Now it needs to get added from command line and using a local .docker.env file (see "how to test" section) 
- edited readme

**Goal**: Set up a continuous building pipeline for the ds4cg inaturalist code base so that when there is a new version of the code it will build two docker images and push them to our docker hub. 

**how to test**: I tested by creating a branch and changing the requirements to build the images when there's a new push to a test repo. 
- test locally: by running instances on local docker app. 
-- `docker compose build` then `docker compose --env-file ./.docker.env up` with local .docker.env file with secrets. tested the UI and made sure APIs still work. 
- tested that pipeline works with new tags :https://github.com/UMassCDS/ds4cg2024-inaturalist/actions/runs/10256595317 & pr: https://github.com/UMassCDS/ds4cg2024-inaturalist/pull/65

dockerhub: https://hub.docker.com/search?q=umasscds

### CODE REVIEW

File: docker-compose.yml [hunk 1]
```diff
@@ -1,7 +1,7 @@
 services:
   frontend:
     platform: linux/amd64
-    image: umasscds/inatator_frontend:0.1.0
+    image: umasscds/inatator_frontend:0.2.0
```
Comment (ginic) [inline docker-compose.yml:4]: I know you updated this for testing purposes, but let's not merge it in until you're ready to 'cut a release'.
```suggestion
    image: umasscds/inatator_frontend:0.1.0
```

File: docker-compose.yml [hunk 2]
```diff
@@ -11,7 +11,7 @@ services:
       - backend
   backend:
     platform: linux/amd64
-    image: umasscds/inatator_backend:0.1.0
+    image: umasscds/inatator_backend:0.2.0
```
Comment (ginic) [inline docker-compose.yml:14]:
```suggestion
    image: umasscds/inatator_backend:0.1.0
```

**Review by ginic — CHANGES_REQUESTED** [review]
Thanks, @angelazhu2 ! The Github workflow looks good and I saw that you tested it and pushed images to Docker Hub. I'm going to delete those v0.2.0 tagged images for now, until we do a proper release on Github.

File: .github/workflows/create-docker-images.yml [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+        uses: docker/setup-qemu-action@v2
+      - name: Set up Docker Buildx
+        id: buildx
+        uses: docker/setup-buildx-action@v2
+      - name: Login to docker hub
+        run: echo "${{ secrets.DOCKER_PASSWORD }}" | docker login -u "${{ secrets.DOCKER_USERNAME }}" --password-stdin
+      - name: 'Create env file'
+        run: |
+          touch .docker.env
+          echo DATABASE_URL=postgresql://${{secrets.POSTGRES_USER}}:${{secrets.POSTGRES_PASSWORD}}@db:5432/${{secrets.POSTGRES_DB}} >> .docker.env
```
Comment (ginic) [inline .github/workflows/create-docker-images.yml:25]: This shouldn't be necessary and it's also not secure. The environment variables can be set using the `--env` or `--env_file` options by the person running the container or via referencing variables in the docker compose file. See https://www.baeldung.com/ops/docker-container-environment-variables or https://www.baeldung.com/ops/docker-compose#managing-environment-variables for more details.

**Review by ginic — CHANGES_REQUESTED** [review]
Thanks, @angelazhu2 ! The `Create env file` step needs to be removed, it's not the appropriate way to configure environment variables in Docker. I don't quite understand why you thought this was needed, so it would probably be helpful to meet and discuss so we can find a better solution.

**Review by cheerstopaula — APPROVED** [review]

File: docker-compose.yml [hunk 4]
```diff
@@ -3,8 +3,11 @@ services:
     container_name: inatator_db
     image: postgres:16
     restart: always
-    env_file:
-      - .docker.env
+    environment:
+      - POSTGRES_DB
```
Comment (SergePogorelov) [inline docker-compose.yml:7]: after changing the `env_file` to these ` environment` variables, do we need to run app and docker compose locally differently?
in other words, do we need to change the instructions in the README file?

**Review by SergePogorelov — CHANGES_REQUESTED** [review]
left a question about README,
and also please don't forget to update CHANGELOG file.

File: .github/workflows/create-docker-images.yml [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v3
+      - name: Set up QEMU
+        uses: docker/setup-qemu-action@v2
+      - name: Set up Docker Buildx
+        id: buildx
+        uses: docker/setup-buildx-action@v2
+      - name: Login to docker hub
+        run: echo "${{ secrets.DOCKER_PASSWORD }}" | docker login -u "${{ secrets.DOCKER_USERNAME }}" --password-stdin
```
Comment (ginic) [inline .github/workflows/create-docker-images.yml:21]: There's a more modern way to do this now, the example I sent you was a little older. It's probably better to use this: 
```suggestion
      - name: Login to Docker Hub
        uses: docker/login-action@v3
        with:
          username: ${{ vars.DOCKERHUB_USERNAME }}
          password: ${{ secrets.DOCKERHUB_TOKEN }}
```

File: .github/workflows/create-docker-images.yml [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+    tags:
+      - 'v*.*.*'
+
+jobs:
+  # Based on instructions from https://www.docker.com/blog/multi-arch-build-and-images-the-simple-way/
+  build-frontend-and-backend:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Checkout code
+        uses: actions/checkout@v3
```
Comment (ginic) [inline .github/workflows/create-docker-images.yml:14]: We can use the most recent version here:
```suggestion
        uses: actions/checkout@v4
```

File: .github/workflows/create-docker-images.yml [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+      - name: Set up QEMU
+        uses: docker/setup-qemu-action@v2
+      - name: Set up Docker Buildx
+        id: buildx
+        uses: docker/setup-buildx-action@v2
```
Comment (ginic) [inline .github/workflows/create-docker-images.yml:19]: These lines are only necessary if we want to do multi-platform builds, which I don't think we do. 
```suggestion
```

File: .github/workflows/create-docker-images.yml [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+      - name: Login to docker hub
+        run: echo "${{ secrets.DOCKER_PASSWORD }}" | docker login -u "${{ secrets.DOCKER_USERNAME }}" --password-stdin
+      - name: Install Docker Compose
+        run: |
+          sudo curl -L "https://github.com/docker/compose/releases/download/$(curl -s https://api.github.com/repos/docker/compose/releases/latest | grep tag_name | cut -d '"' -f 4)/docker-compose-$(uname -s)-$(uname -m)" -o /usr/local/bin/docker-compose
+          sudo chmod +x /usr/local/bin/docker-compose
+          docker-compose --version
```
Comment (ginic) [inline .github/workflows/create-docker-images.yml:26]: This isn't necessary, see my next commend on `docker compose`
```suggestion
```

File: .github/workflows/create-docker-images.yml [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+      - name: Build and push images
+        run: |
+          docker-compose -f docker-compose.yml build
+          docker-compose -f docker-compose.yml push
```
Comment (ginic) [inline .github/workflows/create-docker-images.yml:30]: Docker compose should already be available, it's just that you don't need the hyphen 😅 See https://github.com/orgs/community/discussions/116610 for more details.
```suggestion
          docker compose -f docker-compose.yml build
          docker compose -f docker-compose.yml push
```

File: docker-compose.yml (hunk 4 above)
Comment (ginic) [reply docker-compose.yml:7]: Yes, nice catch, the Readme instructions should be updated as well!

**Review by ginic — CHANGES_REQUESTED** [review]
Thanks for figuring out the environment variable stuff, @angelazhu2! I was able to test the images you pushed to Docker Hub for 0.1.0 and everything worked well. 

There are still some things that need to be addressed:
- The Github action can be simplified a lot. I've put everything in suggestions. 
- The README needs to be updated with the instructions for passing the .env file

Once that's done and this PR is merged, we should: 
- Delete the tags and releases created for testing, both on GitHub and Docker Hub. 
- Finalize the contribution guidelines and release process Google Doc
- Restore the tag for the original v0.1.0 (we'll be able to do this by creating a release/tag for v0.1.0 on the commit that updated the changelog for v0.1.0)
- Release a v1.0.0 version (probably the final version we'll deliver this summer)

File: .github/workflows/create-docker-images.yml (hunk 7 above)
Comment (oz03-hub) [reply .github/workflows/create-docker-images.yml:19]: I am not familiar with the github worklow portion, but we want to target linux/amd64 platform while building, without these lines would it still work?

**Review by ginic — APPROVED** [review]
Looks good! Nice work, @angelazhu2 !

**Pull request merged**